app/drivers/zephyr/battery_voltage_divider.c

```diff
@@ -0,0 +1,214 @@
+    // Magic function that maps mV to this discharge graph from adafruit:
+    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
+    return round(106.818 +
+                 (-0.032685 - 106.818) / pow(1 + pow(bat_mv / 3679.35, 58.979), 0.347386));
+}
+
+static int bvd_sample_fetch(struct device *dev, enum sensor_channel chan) {
+    struct bvd_data *drv_data = dev->driver_data;
+    const struct bvd_config *drv_cfg = dev->config_info;
+    struct adc_sequence *as = &drv_data->as;
+
+    // Make sure selected channel is supported
+    if (chan != SENSOR_CHAN_GAUGE_VOLTAGE && chan != SENSOR_CHAN_GAUGE_STATE_OF_CHARGE) {
+        return -ENOTSUP;
+    }
+
+    int rc = 0;
+
+    // Enable power GPIO if present
+    if (drv_data->gpio) {
+        rc = gpio_pin_set(drv_data->gpio, drv_cfg->power_gpios.pin, 1);
+
+        if (rc != 0) {
+            LOG_DBG("Failed to enable ADC power GPIO: %d", rc);
+            return rc;
+        }
+    }
+
+    // Read ADC
+    rc = adc_read(drv_data->adc, as);
+    as->calibrate = false;
+
+    if (rc == 0) {
+        int32_t val = drv_data->adc_raw;
+
+        adc_raw_to_millivolts(adc_ref_internal(drv_data->adc), drv_data->acc.gain, as->resolution,
+                              &val);
+
+        uint16_t millivolts = val * (uint64_t)drv_cfg->full_ohm / drv_cfg->output_ohm;
+        LOG_DBG("ADC raw %d ~ %d mV => %d mV\n", drv_data->adc_raw, val, millivolts);
+        uint8_t percent = lithium_ion_mv_to_pct(millivolts);
+        LOG_DBG("Percent: %d", percent);
+
+        drv_data->voltage = millivolts;
+        drv_data->state_of_charge = percent;
+    } else {
+        LOG_DBG("Failed to read ADC: %d", rc);
+    }
+
+    // Disable power GPIO if present
+    if (drv_data->gpio) {
+        rc = gpio_pin_set(drv_data->gpio, drv_cfg->power_gpios.pin, 0);
+
+        if (rc != 0) {
+            LOG_DBG("Failed to disable ADC power GPIO: %d", rc);
+        }
+    }
+
+    return rc;
+}
+
+static int bvd_channel_get(struct device *dev, enum sensor_channel chan, struct sensor_value *val) {
+    struct bvd_data *drv_data = dev->driver_data;
+
+    switch (chan) {
+    case SENSOR_CHAN_GAUGE_VOLTAGE:
+        val->val1 = drv_data->voltage / 1000;
+        val->val2 = (drv_data->voltage % 1000) * 1000U;
+        break;
+
+    case SENSOR_CHAN_GAUGE_STATE_OF_CHARGE:
+        val->val1 = drv_data->state_of_charge;
+        val->val2 = 0;
+        break;
+
+    default:
+        return -ENOTSUP;
+    }
+
+    return 0;
+}
+
+static const struct sensor_driver_api bvd_api = {
+    .sample_fetch = bvd_sample_fetch,
+    .channel_get = bvd_channel_get,
+};
+
+static int bvd_init(struct device *dev) {
+    struct bvd_data *drv_data = dev->driver_data;
+    const struct bvd_config *drv_cfg = dev->config_info;
+
+    drv_data->adc = device_get_binding(drv_cfg->io_channel.label);
+
+    if (drv_data->adc == NULL) {
+        LOG_ERR("ADC %s failed to retrieve", drv_cfg->io_channel.label);
+        return -ENOENT;
```

Comment:
Minor: the convention is usually to return `-ENODEV` when a `device_get_binding` returns `NULL`.
(feel free to ignore :) )